lib/Frontend/nvqpp/ConvertExpr.cpp

```diff
@@ -2189,28 +2187,48 @@ bool QuakeBridgeVisitor::VisitInitListExpr(clang::InitListExpr *x) {
     return pushValue(last[0]);
   }
 
-  // Let's allocate some memory and store the init list elements there.
-  // Add the array size value
-  Value arrSize =
-      getConstantInt(builder, loc, last.size(), builder.getI64Type());
+  // These initializer expressions are not a quantum references. In this case,
+  // we allocate some memory for a variable and store the init list elements
+  // there. Add the array size value
+  auto structTy = dyn_cast<cc::StructType>(initListTy);
+  std::int32_t structMems = structTy ? structTy.getMembers().size() : 0;
+  std::int32_t numEles = structMems ? size / structMems : size;
+  Value arrSize = builder.create<arith::ConstantIntOp>(loc, numEles, 64);
 
   // Allocate the required memory chunk.
-  // TODO: If the type is correctly constructed here it will be something like
-  // !cc.array<i32 x 4>. If that is the case, then we can drop the creation of
-  // the `arrSize` value altogether here. The size must be a compile-time
-  // constant as this is coming from a semantic-form of an initializer-list.
   Type eleTy = [&]() {
     if (auto arrTy = dyn_cast<cc::ArrayType>(initListTy))
       return arrTy.getElementType();
     return initListTy;
   }();
-  Value alloca = builder.create<cc::AllocaOp>(loc, eleTy, arrSize);
+  Value alloca = (numEles > 1)
+                     ? builder.create<cc::AllocaOp>(loc, eleTy, arrSize)
+                     : builder.create<cc::AllocaOp>(loc, eleTy);
 
   // Store the values in the allocated memory
-  for (std::size_t i = 0; auto v : last) {
-    Value ptr = builder.create<cc::ComputePtrOp>(
-        loc, cc::PointerType::get(eleTy), alloca,
-        getConstantInt(builder, loc, i++, builder.getI64Type()));
+  for (auto iter : llvm::enumerate(last)) {
+    std::int32_t i = iter.index();
+    auto v = iter.value();
+    Value ptr;
+    if (structMems) {
+      if (numEles > 1) {
+        auto ptrTy =
+            cc::PointerType::get(structTy.getMembers()[i % structMems]);
+        ptr = builder.create<cc::ComputePtrOp>(
+            loc, ptrTy, alloca,
+            ArrayRef<cc::ComputePtrArg>{i / structMems, i % structMems});
+      } else {
+        auto ptrTy = cc::PointerType::get(structTy.getMembers()[i]);
+        ptr = builder.create<cc::ComputePtrOp>(
+            loc, ptrTy, alloca, ArrayRef<cc::ComputePtrArg>{0, i});
+      }
+    } else {
+      auto ptrTy = cc::PointerType::get(eleTy);
+      ptr = builder.create<cc::ComputePtrOp>(loc, ptrTy, alloca,
+                                             ArrayRef<cc::ComputePtrArg>{i});
+    }
+    assert(ptr && (v.getType() ==
+                   cast<cc::PointerType>(ptr.getType()).getElementType()));
```

Comment:
maybe add a assert message to the end of this

Reply:
[366cb3f](https://github.com/NVIDIA/cuda-quantum/pull/1068/commits/366cb3fd5ad550d3fb37f614041fe7bc4bc58a73)